### PR TITLE
[TON] Support TON tx compiler

### PR DESCRIPTION
**Pull request by 10gic**
## Description

Implement the TON transaction compiler, fix https://github.com/trustwallet/wallet-core/issues/3797

## How to test

Run Rust, C++, Kotlin, Swift tests

## Checklist

- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

**Comment by 10gic** [comment]
Hi @satoshiotomakan, I would greatly appreciate it if you could take the time to review this pull request.

**Comment by satoshiotomakan** [comment]
Hi @10gic, thank you for the PR! I'll review today

File: rust/chains/tw_ton/src/wallet/mod.rs [hunk 1]
```diff
@@ -33,13 +33,6 @@ impl VersionedTonWallet {
         }
     }
 
-    pub fn state_init(&self) -> CellResult<StateInit> {
-        match self {
-            Self::V4R2(wallet_v4r2) => wallet_v4r2.state_init(),
-            Self::V5R1(wallet_v5r1) => wallet_v5r1.state_init(),
-        }
-    }
```
Comment (satoshiotomakan) [inline rust/chains/tw_ton/src/wallet/mod.rs:41]: Why did you remove this method? It's used in `compile_signed_transaction`

File: rust/chains/tw_ton/src/signing_request/builder.rs [hunk 2]
```diff
@@ -46,6 +48,17 @@ impl SigningRequestBuilder {
         })
     }
 
+    /// Returns the cell hash of external message to sign.
+    pub fn msg_to_sign(input: &Proto::SigningInput) -> SigningResult<H256> {
```
Comment (satoshiotomakan) [inline rust/chains/tw_ton/src/signing_request/builder.rs:52]: There is no need in this function as suggested above

File: rust/chains/tw_ton/src/signer.rs [hunk 3]
```diff
@@ -25,35 +22,16 @@ impl TheOpenNetworkSigner {
...
-        let state_init = signing_request.seqno == 0;
-        let signed_tx = signing_request
-            .wallet
-            .sign_transaction(external_message, state_init)
-            .context("Error signing/wrapping an external message")?
-            .build()
-            .context("Error generating signed message cell")
-            .map_err(cell_to_signing_error)?;
+        let msg_to_sign = SigningRequestBuilder::msg_to_sign(&input)?;
+        let wallet = SigningRequestBuilder::build(&input)?.wallet;
```
Comment (satoshiotomakan) [inline rust/chains/tw_ton/src/signer.rs:29]: `SigningRequestBuilder::build` is called three times in `TheOpenNetworkSigner::sign_impl`: 
1. `SigningRequestBuilder::msg_to_sign` calls it to generate a message to sign
2. `SigningRequestBuilder::build` itself to get the wallet only.
3. `TheOpenNetworkCompiler::compile_impl` to get the message to be compiled.

`TheOpenNetworkSigner::sign_impl` can be refactored this way:
1. Call `SigningRequestBuilder::build` one time to get both the signing request and wallet.
2. Create an external message via `ExternalMessageCreator::create_external_message_to_sign`.
3. Sign it using `wallet.sign_transaction()`.

`TheOpenNetworkCompiler::pre_image_hash_impl` will do:
1. Call `SigningRequestBuilder::build` one time to get both the signing request.
2. Create an external message via `ExternalMessageCreator::create_external_message_to_sign`.
3. Return the external message hash.

And `TheOpenNetworkCompiler::compile_impl` will do:
1. Call `SigningRequestBuilder::build` one time to get both the signing request.
2. Create an external message via `ExternalMessageCreator::create_external_message_to_sign`.
3. Compile it using `wallet.compile_transaction()`.

**Review by satoshiotomakan — CHANGES_REQUESTED** [review]
Good job again. Please consider making changes to avoid ambiguous API and doing same work multiple times

File: rust/chains/tw_ton/src/signer.rs (hunk 3 above)
Comment (10gic) [reply rust/chains/tw_ton/src/signer.rs:29]: My initial idea was to have the sign_impl code reuse as much code from other parts as possible to reduce redundancy. Your approach results in some duplicate code between sign_impl and compile_impl, but it's also very good, so I adopted your method.

**Comment by 10gic** [comment]
> Good job again. Please consider making changes to avoid ambiguous API and doing same work multiple times

Thank you very much for your comments. I've adjusted the code according to your advice. Could you please review it again?

File: rust/chains/tw_ton/src/signer.rs (hunk 3 above)
Comment (satoshiotomakan) [reply rust/chains/tw_ton/src/signer.rs]: Yes that makes sense of course, but there is a bit duplication that makes the code faster, and uses less memory allocations

**Review by satoshiotomakan — APPROVED** [review]
LGTM

**Comment by 10gic** [comment]
Hi @satoshiotomakan. I'm wondering if it's a good time to proceed with merging the code. Please let me know if there's anything else that needs to be done beforehand.

**Comment by enriquesouza** [comment]
@10gic thank you a lot.

**Review by KATERYNA3589 — COMMENTED** [review]
Thanks
